iree/turbine/kernel/wave/codegen/read_write.py

```diff
@@ -421,33 +438,75 @@ def _create_vec_read(
         data, offset_th = _linearize_memref(
             mem, start_indices_wg, start_indices_th, strides
         )
-        offset_th = vector_d.splat(offsets_vec.type, offset_th)
-        offsets_vec = arith_d.addi(offsets_vec, offset_th)
-        if mask is not None:
+
+        if offsets_vec is None:
+            offsets_vec_type = VectorType.get(vector_type.shape, IndexType.get())
+            vals = [
+                IntegerAttr.get(IndexType.get(), v) for v in range(elements_per_thread)
+            ]
+            offsets_vec = arith_d.constant(
+                offsets_vec_type, DenseElementsAttr.get(vals, offsets_vec_type)
+            )
+
+        if splatted_masked:
+            # If mask value is same for all vector elements, we can use vector
+            # buffer ops.
             i32 = IntegerType.get_signless(32)
-            i32vec = VectorType.get([elements_per_thread], i32)
-            offsets_vec = arith_d.index_cast(i32vec, offsets_vec)
+            offset_th = arith_d.index_cast(i32, offset_th)
             oob_idx = _get_max_buffer_size(element_type)
             oob_idx = arith_d.constant(i32, oob_idx)
-            oob_idx = vector_d.splat(offsets_vec.type, oob_idx)
-            offsets_vec = arith_d.select(mask, offsets_vec, oob_idx)
+            offset_th = arith_d.select(mask_splat, offset_th, oob_idx)
 
-        for i in range(elements_per_thread):
-            offset = vector_d.extract(
-                offsets_vec, static_position=[i], dynamic_position=[]
-            )
-            if mask is None:
-                elem = memref_d.load(element_type, data, indices=[offset])
-            else:
-                elem = amdgpu_d.raw_buffer_load(element_type, data, indices=[offset])
+            # Buffer ops doesn't support 1-element vectors, convert to scalar.
+            load_type = vector_type
+            if elements_per_thread == 1:
+                load_type = element_type
 
-            result = vector_d.insert(
-                elem, result, static_position=[i], dynamic_position=[]
-            )
+            res = amdgpu_d.raw_buffer_load(load_type, data, indices=[offset_th])
+            if elements_per_thread == 1:
+                res = vector_d.splat(vector_type, res)
 
-        return result
+            return res
+        else:
+            # If mask value is different for each element, unroll op to
+            # individual values.
+            offset_th = vector_d.splat(offsets_vec.type, offset_th)

```

Comment:
These functions are almost identical for read/write, can we refactor them into a common function?

Reply:
done